### DISCUSSION TITLE
Is it possible to replace default text editor with a better one?

### DISCUSSION THREAD

**Discussion by nashv** [datenstrom/community#103] [Ask a question] [answered]
The default text editor is sufficient but hardly the best. How difficult/easy would it be to intrgrate something like EpicEditor (http://epiceditor.com) into Yellow ?

**Comment by markseuffert** [comment]
Good idea to add another text editor. Can you describe what features you are missing most? Maybe someone has suggestions or wants to write a plugin? Any help would be much appreciated. :smile:

**Comment by nashv** [comment]
I missing the ability to save drafts without publishing using the Drafts plugin.  And simple formatting buttons/post preview.
I could write a plugin, but was looking for pointers on exactly how one would start. I am not too familiar with the Yellow source code.

**Comment by i-do** [comment]
Yes, I think it's good idea. I'd like to see [ZenPen](https://github.com/tholman/zenpen) as editor. It so clear and simple!

**Comment by nibreh** [comment]
HI - I use [Markbar by reinink](https://github.com/reinink/jQuery.Markbar)

I made a repository on how to install in Yellow:
https://github.com/nibreh/yellow-plugin-markbar

![](https://cloud.githubusercontent.com/assets/7314006/9836901/5bedad82-5a2c-11e5-8619-35bc6428eb28.png)

It's very usefull

**Comment by i-do** [comment]
Wow! Look on [this markdown editor](http://tibastral.github.io/markdownify/). It very simple, have preview and can upload files.

**Comment by nibreh** [comment]
Hey mardownify is really cool. But I don't know how to use it in Yellow, like many others editor.

Markbar is the only one that works with panel edit from yellow.

**Comment by markseuffert** [comment]
If someone would like to see this feature, please add +1 to the top of this issue.

**Comment by BsNoSi** [comment]
Markdonify looks really nice. I love the "preview" option and the smart presentation of markdown. But I miss a way for uploading images or files in general. The formatting bar is something I do not need, because for me this as an antagonism to markdown ;)

**Comment by i-do** [comment]
+1

**Comment by redtide** [comment]
It could be nice if the webinterface plugin could be splitted (e.g. in 2 'user-admin' and 'simple-editor' plugins), so it could be easier to replace the editor with a custom one (e.g. I would like to use simplemde on a new custom plugin) without removing the current one.

**Comment by redtide** [comment]
Here some start points; the following plugins are currently non-working / broken / buggy so feel free to clone / modify / fix stuff / send pull requests:
- [Bootstrap Markdown](https://github.com/redtide/yellow-bootstrap-markdown-plugin)
- [Markdownify](https://github.com/redtide/yellow-markdownify-plugin) (suggested by @i-do)
- [SimpleMDE](https://github.com/redtide/yellow-simplemde-plugin) (based on [Lepture](https://github.com/lepture/editor) suggested by @nibreh)
- [Zenpen](https://github.com/redtide/yellow-zenpen-plugin) (suggested by @i-do)

Some issues with these plugins seems related to the fact that some interactions with the webinterface plugin are needed (e.g. for resize/refresh) to get updated in sync with the yellow-pane-edit-page textarea.

**Comment by markseuffert** [comment]
I am working on this now.

You can see what's already working in the [Yellow demo](https://datenstrom.se/yellow/demo/). There's a new editor with formatting toolbar. It's based on your feedback and I tried to integrate what features editors have in common. The formatting toolbar was difficult to do, it went trough many design reviews and cups of coffee, until it was no longer terrible to use. The preview feature was surprisingly little work, just a XMLHttpRequest and the server will run it trough the Markdown parser. I will upload the source code after some more testing. 

Let me know if you have questions or feedback.

**Comment by schulle4u** [comment]
Great work Mark! From a blind user's perspective I can say that this is one of the best online editing solutions I've seen so far. I never was a fan of WYSIWYG editors or formating toolbars as they often cause trouble for screenreaders, but your solution keeps the old editor working as usual, without breaking anything. It would be great to have some keyboard shortcuts for even easier text formating. 

As for the "Preview" button, I suggest to label it as "Toggle preview" for clarification and to avoid confusion after someone hits preview and is wondering how to return to the text input. 

Steffen

**Comment by schulle4u** [comment]
@markseu That was quick, thanks for adding keyboard shortcuts. The default shortcuts however are not usable in most browsers (ctrl+b opens bookmarks etc). I suggest to add an additional modifier key to the default values. Ctrl+alt should work in most cases, just tested this in two browsers without problems.

**Comment by markseuffert** [comment]
Doesn't GitHub use the same keyboard shortcuts, e.g. for ctrl+b (or cmd+b) for bold?

**Comment by schulle4u** [comment]
Hmm indeed. Thats interesting, Github's markdown shortcuts don't collide with the browser shortcuts, but I have no clue why this could happen in YellowEdit. Probably a nasty JS issue.

**Comment by markseuffert** [comment, marked as answer]
The new editor is available now.

Here's what has been changed:

* formatting buttons, you can configure different toolbar buttons
* keyboard shortcuts, you can configure your own keyboard shortcuts
* button for page preview, toggles between editor and preview
* button for draft status, if the draft plugin is installed
* list of icons, if the fontawesome plugin is installed
* list of emoji, if the emojiawesome plugin is installed

Default keyboard shortcut | Description
-- | --
ctrl+b or cmd+b | Markdown formatting for bold
ctrl+i or cmd+i | Markdown formatting for italic
ctrl+e or cmd+e | Markdown formatting for code
ctrl+k or cmd+k | Markdown formatting for link
ctrl+s or cmd+s | Save page
ctrl+shift+p or cmd+shift+p | Preview  page



Thanks for all your feedback. Let me know how this works for you.
